### ISSUE TITLE
xkb::SelectEvents is wrong

### ISSUE THREAD

**Issue by eduardosm** [psychon/x11rb#281]
The request function is generated as:
```rust
pub fn select_events<'c, Conn>(conn: &'c Conn, device_spec: DeviceSpec, clear: u16, select_all: u16, affect_map: u16, map: u16, details: &SelectEventsAux) -> Result<VoidCookie<'c, Conn>, ConnectionError>
where Conn: RequestConnection + ?Sized
{
    let extension_information = conn.extension_information(X11_EXTENSION_NAME)?
        .ok_or(ConnectionError::UnsupportedExtension)?;
    let affect_which = details.value_mask();
    let details_bytes = details.serialize();
    let length: usize = (16 + details_bytes.len() + 3) / 4;
    let length_bytes = TryInto::<u16>::try_into(length).unwrap_or(0).serialize();
    let device_spec_bytes = device_spec.serialize();
    let affect_which_bytes = affect_which.serialize();
    let clear_bytes = clear.serialize();
    let select_all_bytes = select_all.serialize();
    let affect_map_bytes = affect_map.serialize();
    let map_bytes = map.serialize();
    let request0 = [
        extension_information.major_opcode,
        SELECT_EVENTS_REQUEST,
        length_bytes[0],
        length_bytes[1],
        device_spec_bytes[0],
        device_spec_bytes[1],
        affect_which_bytes[0],
        affect_which_bytes[1],
        clear_bytes[0],
        clear_bytes[1],
        select_all_bytes[0],
        select_all_bytes[1],
        affect_map_bytes[0],
        affect_map_bytes[1],
        map_bytes[0],
        map_bytes[1],
    ];
    let length_so_far = (&request0).len();
    let length_so_far = length_so_far + (&details_bytes).len();
    let padding1 = &[0; 3][..(4 - (length_so_far % 4)) % 4];
    let length_so_far = length_so_far + (&padding1).len();
    assert_eq!(length_so_far, length * 4);
    Ok(conn.send_request_without_reply(&[IoSlice::new(&request0), IoSlice::new(&details_bytes), IoSlice::new(&padding1)], Vec::new())?)
}
```

But the XML definition is:
```xml
<request name="SelectEvents" opcode="1">
	<field name="deviceSpec" type="DeviceSpec" />
	<field name="affectWhich" type="CARD16" mask="EventType" />
	<field name="clear" type="CARD16" mask="EventType" />
	<field name="selectAll" type="CARD16" mask="EventType" />
	<field name="affectMap" type="CARD16" mask="MapPart" />
	<field name="map" type="CARD16" mask="MapPart" />
	<switch name="details">
		<op op="&amp;">
			<fieldref>affectWhich</fieldref>
			<op op="&amp;">
				<unop op="~"><fieldref>clear</fieldref></unop>
				<unop op="~"><fieldref>selectAll</fieldref></unop>
			</op>
		</op>
		<bitcase>
			<enumref ref="EventType">NewKeyboardNotify</enumref>
			<field name="affectNewKeyboard" type="CARD16" mask="NKNDetail" />
			<field name="newKeyboardDetails" type="CARD16" mask="NKNDetail" />
		</bitcase>
		<bitcase>
			<enumref ref="EventType">StateNotify</enumref>
			<field name="affectState" type="CARD16" mask="StatePart" />
			<field name="stateDetails" type="CARD16" mask="StatePart" />
		</bitcase>
		<bitcase>
			<enumref ref="EventType">ControlsNotify</enumref>
			<field name="affectCtrls" type="CARD32" mask="Control" />
			<field name="ctrlDetails" type="CARD32" mask="Control" />
		</bitcase>
		<bitcase>
			<enumref ref="EventType">IndicatorStateNotify</enumref>
			<field name="affectIndicatorState" type="CARD32" />
			<field name="indicatorStateDetails" type="CARD32" />
		</bitcase>
		<bitcase>
			<enumref ref="EventType">IndicatorMapNotify</enumref>
			<field name="affectIndicatorMap" type="CARD32" />
			<field name="indicatorMapDetails" type="CARD32" />
		</bitcase>
		<bitcase>
			<enumref ref="EventType">NamesNotify</enumref>
			<field name="affectNames" type="CARD16" mask="NameDetail" />
			<field name="namesDetails" type="CARD16" mask="NameDetail" />
		</bitcase>
		<bitcase>
			<enumref ref="EventType">CompatMapNotify</enumref>
			<field name="affectCompat" type="CARD8" mask="CMDetail" />
			<field name="compatDetails" type="CARD8" mask="CMDetail" />
		</bitcase>
		<bitcase>
			<enumref ref="EventType">BellNotify</enumref>
			<field name="affectBell" type="CARD8" />
			<field name="bellDetails" type="CARD8" />
		</bitcase>
		<bitcase>
			<enumref ref="EventType">ActionMessage</enumref>
			<field name="affectMsgDetails" type="CARD8" />
			<field name="msgDetails" type="CARD8" />
		</bitcase>
		<bitcase>
			<enumref ref="EventType">AccessXNotify</enumref>
			<field name="affectAccessX" type="CARD16" mask="AXNDetail" />
			<field name="accessXDetails" type="CARD16" mask="AXNDetail" />
		</bitcase>
		<bitcase>
			<enumref ref="EventType">ExtensionDeviceNotify</enumref>
			<field name="affectExtDev" type="CARD16" mask="XIFeature" />
			<field name="extdevDetails" type="CARD16" mask="XIFeature" />
		</bitcase>
	</switch>
</request>
```

More specifically:
```xml
<op op="&amp;">
	<fieldref>affectWhich</fieldref>
	<op op="&amp;">
		<unop op="~"><fieldref>clear</fieldref></unop>
		<unop op="~"><fieldref>selectAll</fieldref></unop>
	</op>
</op>
```

The `affectWhich` field value cannot be deduced only from `details`.

**Comment by psychon** [comment]
Sigh, I feel like more than half the bugs in the code generator happen because something ignores an `.expr` field when it shouldn't (or mishandles one).

How did you find this? Are you trying to actually use xkb?

**Comment by eduardosm** [comment]
> How did you find this? Are you trying to actually use xkb?

Yes, I was replacing some `unsafe`-FFI with x11rb, that's how I found this and https://github.com/psychon/x11rb/pull/282.

**Comment by psychon** [comment]
So... how should this look like? From a quick look at libX11, there is `XkbSelectEvents()` which does some magic and looks like a convenience function: https://sources.debian.org/src/libx11/2:1.6.9-2/src/xkb/XKB.c/?hl=84#L84

There is also `XkBSelectEventDetails` which seems to be the more complicated cousin: https://sources.debian.org/src/libx11/2:1.6.9-2/src/xkb/XKB.c/?hl=164#L164
(But has a similar prototype?!?)
Looking at what the code actually does... meh, this does not help the issue here.

I guess we should just add `affectWhich` as an argument to the function and then `assert!` that the expression that has to hold actually holds? (Or perhaps, instead of `assert!`, return an `Err`). Is there some better solution than this?

**Comment by eduardosm** [comment]
> I guess we should just add `affectWhich` as an argument to the function and then `assert!` that the expression that has to hold actually holds? (Or perhaps, instead of `assert!`, return an `Err`). Is there some better solution than this?

I would do that, other solutions would be either too hard or manually write the function. If the expression does not hold, it's the programmer's fault, so I would use an `assert!`.

**Comment by psychon** [comment]
Possible patch:
```diff
diff --git a/code_generator_helpers/request.py b/code_generator_helpers/request.py
index b7bf1a3..ed3575f 100644
--- a/code_generator_helpers/request.py
+++ b/code_generator_helpers/request.py
@@ -38,14 +38,23 @@ def handle_switches(module, obj, name, function_name):
     switch = switches[0]
 
     # Find the mask field for the switch
-    lenfield_name = switch.type.expr.lenfield_name
-    mask_field = list(filter(lambda field: field.field_name == lenfield_name, obj.fields))
-    assert len(mask_field) == 1
-    mask_field = mask_field[0]
+    expr = switch.type.expr
+    if expr.op is None:
+        lenfield_name = expr.lenfield_name
+        mask_field = list(filter(lambda field: field.field_name == lenfield_name, obj.fields))
+        assert len(mask_field) == 1
+        mask_field = mask_field[0]
+
+        # Hide it from the API and "connect" it to the switch
+        mask_field.visible = False
+        mask_field.lenfield_for_switch = switch
+    else:
+        # Well, this case is complicated. We handle it as a special case.
+        from collections import namedtuple
+        from xcbgen.xtypes import tcard16
 
-    # Hide it from the API and "connect" it to the switch
-    mask_field.visible = False
-    mask_field.lenfield_for_switch = switch
+        assert name == ('xcb', 'xkb', 'SelectEvents')
+        mask_field = namedtuple('FakeField', 'type field_type')(tcard16, ('uint16_t',))
 
     module._generate_aux(aux_name, obj, switch, mask_field, function_name)
     return aux_name
@@ -337,6 +346,11 @@ def request_implementation(module, obj, name, fds, fds_is_list):
                     module.out("let %s = %s.serialize();", field_bytes, source)
             if field.type.is_switch or field.type.size is None:
                 # We have a byte array that we can directly send
+                if field.type.is_switch and field.type.expr.op is not None:
+                    expr = module.expr_to_str(field.type.expr)
+                    msg = "TODO: helpful message here"
+                    module.out("assert_eq!(%s, %s.value_mask(), \"%s\");",
+                               expr, rust_variable, msg)
                 _emit_request()
                 _emit_add_to_requests("&" + field_bytes)
             else:
```
Effect on the generated code:
```diff
diff -Nurp "target/code-gen-ref" "target/code-gen-cmp"
diff -Nurp target/code-gen-ref/xkb.rs target/code-gen-cmp/xkb.rs
--- target/code-gen-ref/xkb.rs	2020-03-10 20:59:32.412804148 +0100
+++ target/code-gen-cmp/xkb.rs	2020-03-10 21:14:21.121452738 +0100
@@ -7385,12 +7385,11 @@ impl Serialize for SelectEventsAux {
         }
     }
 }
-pub fn select_events<'c, Conn>(conn: &'c Conn, device_spec: DeviceSpec, clear: u16, select_all: u16, affect_map: u16, map: u16, details: &SelectEventsAux) -> Result<VoidCookie<'c, Conn>, ConnectionError>
+pub fn select_events<'c, Conn>(conn: &'c Conn, device_spec: DeviceSpec, affect_which: u16, clear: u16, select_all: u16, affect_map: u16, map: u16, details: &SelectEventsAux) -> Result<VoidCookie<'c, Conn>, ConnectionError>
 where Conn: RequestConnection + ?Sized
 {
     let extension_information = conn.extension_information(X11_EXTENSION_NAME)?
         .ok_or(ConnectionError::UnsupportedExtension)?;
-    let affect_which = details.value_mask();
     let details_bytes = details.serialize();
     let length: usize = (16 + details_bytes.len() + 3) / 4;
     let length_bytes = TryInto::<u16>::try_into(length).unwrap_or(0).serialize();
@@ -7400,6 +7399,7 @@ where Conn: RequestConnection + ?Sized
     let select_all_bytes = select_all.serialize();
     let affect_map_bytes = affect_map.serialize();
     let map_bytes = map.serialize();
+    assert_eq!((affect_which) & ((!(clear)) & (!(select_all))), details.value_mask(), "TODO: helpful message here");
     let request0 = [
         extension_information.major_opcode,
         SELECT_EVENTS_REQUEST,
@@ -11014,9 +11014,9 @@ pub trait ConnectionExt: RequestConnecti
         use_extension(self, wanted_major, wanted_minor)
     }
 
-    fn xkb_select_events<'c>(&'c self, device_spec: DeviceSpec, clear: u16, select_all: u16, affect_map: u16, map: u16, details: &SelectEventsAux) -> Result<VoidCookie<'c, Self>, ConnectionError>
+    fn xkb_select_events<'c>(&'c self, device_spec: DeviceSpec, affect_which: u16, clear: u16, select_all: u16, affect_map: u16, map: u16, details: &SelectEventsAux) -> Result<VoidCookie<'c, Self>, ConnectionError>
     {
-        select_events(self, device_spec, clear, select_all, affect_map, map, details)
+        select_events(self, device_spec, affect_which, clear, select_all, affect_map, map, details)
     }
 
     fn xkb_bell(&self, device_spec: DeviceSpec, bell_class: BellClassSpec, bell_id: IDSpec, percent: i8, force_sound: bool, event_only: bool, pitch: i16, duration: i16, name: ATOM, window: WINDOW) -> Result<VoidCookie<'_, Self>, ConnectionError>
```
@eduardosm Could you check if this is actually correct? Also, do you have any suggestions for "TODO: helpful message here"? It seems... complicated.

**Comment by psychon** [comment]
I went through `git grep expr.lenfield_name`. The other uses look correct (they check `expr.op` before just using `lenfield_name`, this "try to propagate `lenfield_name` up"-behaviour in xcbgen has been a constant source of bugs).

**Comment by eduardosm** [comment]
> ```diff
> +        assert name == ('xcb', 'xkb', 'SelectEvents')
> +        mask_field = namedtuple('FakeField', 'type field_type')(tcard16, ('uint16_t',))
> ```

Looks a bit ad-hoc, but the assert prevents from incorrectly handling other cases.

I have tested it and seems to work, both for the panicking and the non-panicking case. 

> "TODO: helpful message here"?

What about something like `msg = "the value passed to '%s' is not compatible with the values of the other request fields" % field.field_name`

**Comment by psychon** [comment]
> Looks a bit ad-hoc, but the assert prevents from incorrectly handling other cases.

Well, it is ad-hoc. This is the only case in `xcb-proto` that `<switch>`es on a (non-trivial) expression. Instead of trying to get the type of all the fields that appear in the expression and asserting that they are all the same (after which I still need some "magic" to use as `mask_field`), I went with this simpler and not-really-that-much-uglier solution.

There is pre-existing art: xproto's `QueryTextExtents` is also a special case with an assert. ;-)
Oh and `DeviceTimeCoord` has an `assert` that then assumes type `u8` for something that I didn't manage to figure out any other way.

Thanks for testing this. I opened #286 with basically the above patch, but a more complicated/helpful message for the `assert_eq!`.
